.github/scripts/simple-smoke-test.sh

```diff
@@ -1,49 +1,88 @@
-      echo "Removing existing '$container_name' container"
-      docker container rm --force "$container_name"
+    remove_container_if_exists "${container_name}"
+
+    check_distribution_type "${image}" "${expected_distribution_type}"
+    check_version "${image}" "${expected_version}"
+
+    test_map_read_write "${image}" "${container_name}"
+}
+
+function remove_container_if_exists() {
+    local container_name=$1
+
+    local containers
+    containers=$(docker ps --all --quiet --filter name="${container_name}")
+
+    if [[ -n ${containers} ]]; then
+      echo "Removing existing '${container_name}' container"
+      docker container rm --force "${container_name}"
     fi
+}
 
-    echo "Checking if $image is EE"
-    if docker run --rm $image bash -c 'compgen -G lib/*enterprise*'; then
+function check_distribution_type() {
+    local image=$1
+    local expected_distribution_type=$2
+
+    echo "Checking ${image} distribution type"
+    if docker run --rm "${image}" bash -c 'compgen -G lib/*enterprise*'; then
       echo "EE contents identified"
       distribution_type="ee"
     else
       echo "No EE contents identified - assuming OSS"
       distribution_type="oss"
     fi
 
-    if [[ "$distribution_type" != "$expected_distribution_type" ]]; then
-      echo "Distribution was $distribution_type, not $expected_distribution_type as expected"
+    if [[ "${distribution_type}" != "${expected_distribution_type}" ]]; then
+      echo "Image ${image} should contain ${expected_distribution_type} distribution but ${distribution_type} was detected"
       exit 1
     fi
+}
+
+function check_version() {
+    local image=$1
+    local expected_version=$2
+
+    echo "Checking ${image} version"
+    local version
+    version=$(docker run --rm "${image}" bin/hz-cli --version | awk '/Hazelcast/ {print $2}')
+    if [[ "${version}" == "${expected_version}" ]]; then
+      echo "${image} version identified as ${version}"
+    else
+      echo "${image} version was ${version}, not ${expected_version} as expected"
+      exit 1
+    fi
+}
+
+function test_map_read_write() {
+    local image=$1
+    local expected_distribution_type=$2
 
-    echo "Starting container '$container_name' from image '$image'"
-    docker run -it --name "$container_name" -e HZ_LICENSEKEY -e HZ_INSTANCETRACKING_FILENAME -d -p5701:5701 "$image"
+    echo "Starting container '${container_name}' from image '${image}'"
+    docker run -it --name "${container_name}" -e HZ_LICENSEKEY -e HZ_INSTANCETRACKING_FILENAME -d -p5701:5701 "${image}"
     local key="some-key"
     local expected="some-value"
-    echo "Putting value '$expected' for key '$key'"
-    while ! clc --timeout 5s map set -n some-map $key $expected --log.path stderr
+    echo "Putting value '${expected}' for key '${key}'"
+    while ! clc --timeout 5s map set -n some-map "${key}" "${expected}" --log.path stderr
     do
       echo "Retrying..."
       sleep 3
```

Comment:
If it fails then we should bail out I think?
don't rely on below get check. it will be clearer where the issue lies IMO

Reply:
> If it fails then we should bail out I think?
don't rely on below get check. it will be clearer where the issue lies IMO

Are you saying we shouldn't poll/retry? Or something else?

Reply:
we should retry but should bail out after some retry count?

Reply:
> we should retry but should bail out after some retry count?

I don’t disagree but this isn’t new for this PR. The action has a timeout anyway (albeit 6 hours I think…)

Reply:
ok. i will resolve